File: geometric/shared part of parabolas.py
from math import sqrt
class QuadraticFunction:
    a = b = c = None

    def __init__(self, a: int, b: int, c: int):
        self.a = a
        self.b = b
        self.c = c

    def calculate(self, x: float):
        return x*x*self.a + x*self.b + self.c

    def get_x_of_common_points_with(self, f):
        a = self.a - f.a  # self.a, f.a != 0 and self.a != f.a 
        b = self.b - f.b
        c = self.c - f.c

        delta = b*b - ((a*c) << 2)

        if not (delta > 0) or c == 0:
            return None, None

        sqrt_of_delta = sqrt(delta)

        root1 = (-b - sqrt_of_delta) / (a << 1)  # a != 0
        root2 = (-b + sqrt_of_delta) / (a << 1)  # a != 0
        if root1 < root2:
            return root1, root2
        return root2, root1

    def calculate_intersection_with(self, f):
        #   a = left side of a trapeze
        #   b = right side of a trapeze
        #   step = height of trapeze
        field = 0

        # Increase n_of_trapezes to increase precission
        n_of_trapezes = 2000  # Can't euqal 0 !
        common_x1, common_x2 = self.get_x_of_common_points_with(f)

        if common_x1 is None:
            return 0

        step = abs(common_x1 - common_x2) / n_of_trapezes
        
        x = common_x1 + step * n_of_trapezes

        y1 = self.calculate(x)
        y2 = f.calculate(x)

        a = abs(y1 - y2)

        while n_of_trapezes:
            b = a

            x -= step

            y1 = self.calculate(x)
            y2 = f.calculate(x)

            a = abs(y1 - y2)

            field += (a + b) / 2 * step #   step never equal 0
            
            n_of_trapezes -= 1
           
        return field

#   Number of tests
n = int(input())

while n:
    #   ax^2 + bx + c
    a, b, c = [int(x) for x in input().split()] 
    f1 = QuadraticFunction(a, b, c)
    
    #   ax^2 + bx + c
    a, b, c = [int(x) for x in input().split()] 
    f2 = QuadraticFunction(a, b, c)

    field = f1.calculate_intersection_with(f2)


    print("{:.2f}".format(round(field, 2)))

    n -= 1

File: geometric/shared part of parabolas.cpp
#include <iostream>
#include <cmath>

using namespace std;

//  Change n_of_trapezes to higher value to increase precission

struct Roots {
    const double FIRST = 0;
    const double SECOND = 0;
    const bool VALID = false;

    Roots(const double FIRST, const double SECOND, const bool VALID)
        :FIRST(FIRST), SECOND(SECOND), VALID(VALID)
    {}

    Roots() {
        // valid = false;
    }
};

struct QuadraticFunction
{
    const int32_t A;
    const int32_t B;
    const int32_t C;

    QuadraticFunction(const int32_t A, const int32_t B, const int32_t C) :
        A(A), B(B), C(C) {}

    double calculate(const double& X) {
        return X * X * this->A + X * this->B + this->C;
    }

    Roots get_x_of_common_points_with(QuadraticFunction& f) {
        const int32_t A = this->A - f.A;
        const int32_t B = this->B - f.B;
        const int32_t C = this->C - f.C;

        const int32_t DELTA = B * B - 4 * A * C;
        if (!(DELTA > 0))
            return Roots(0, 0, false);

        const double SQRT_OF_DELTA = sqrt(DELTA);
        const int64_t TWO_A = int64_t(A) * 2;
        const double ROOT1 = (-B - SQRT_OF_DELTA) / TWO_A;
        const double ROOT2 = (-B + SQRT_OF_DELTA) / TWO_A;

        if (ROOT1 < ROOT2)
            return Roots(ROOT1, ROOT2, true);
        return Roots(ROOT2, ROOT1, true);
    }

    double calculate_intersection_with(QuadraticFunction& f) {
        double field = 0;

        // Can't euqal 0 !
        uint64_t n_of_trapezes = 3000;

        const Roots COMMONS = this->get_x_of_common_points_with(f);
        if (!COMMONS.VALID)
            return 0;

        const double STEP = abs(COMMONS.FIRST - COMMONS.SECOND) / n_of_trapezes;

        double x = COMMONS.FIRST + STEP * n_of_trapezes;

        double y1 = this->calculate(x);
        double y2 = f.calculate(x);

        double a = abs(y1 - y2);
        double b;

        const double TRAPEZE_HEIGHT_DIVIDED_BY_TWO = STEP / 2;

        while (n_of_trapezes--) {
            b = a;
            x -= STEP;

            y1 = this->calculate(x);
            y2 = f.calculate(x);

            a = abs(y1 - y2);

            field += (a + b) * TRAPEZE_HEIGHT_DIVIDED_BY_TWO; //  step never equal 0
        }
        return field;
    }
};

int main()
{
    uint64_t n;
    cin >> n;
    double result;

    while (n--) {
        int32_t a, b, c;
        cin >> a >> b >> c;
        QuadraticFunction f1(a, b, c);
        cin >> a >> b >> c;
        QuadraticFunction f2(a, b, c);

        result = f1.calculate_intersection_with(f2);

        printf("%.2f\n", float(uint64_t(round(result * 100)) / 100.0));
    }




}
